test/core/surface/channel_init_test.cc

```diff
@@ -0,0 +1,172 @@
+}
+
+TEST(ChannelInitTest, BeforeConstraintsApply) {
+  ChannelInit::Builder b;
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("foo"))
+      .Before({FilterNamed("bar")});
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("bar"));
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("baz"));
+  auto init = b.Build();
+  EXPECT_EQ(GetFilterNames(init, GRPC_CLIENT_CHANNEL, ChannelArgs()),
+            std::vector<std::string>({"baz", "foo", "bar"}));
+}
+
+TEST(ChannelInitTest, PredicatesCanFilter) {
+  ChannelInit::Builder b;
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("foo"))
+      .IfChannelArg("foo", true);
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("bar"))
+      .IfChannelArg("bar", false);
+  auto init = b.Build();
+  EXPECT_EQ(GetFilterNames(init, GRPC_CLIENT_CHANNEL, ChannelArgs()),
+            std::vector<std::string>({"foo"}));
+  EXPECT_EQ(GetFilterNames(init, GRPC_CLIENT_CHANNEL,
+                           ChannelArgs().Set("foo", false)),
+            std::vector<std::string>({}));
+  EXPECT_EQ(
+      GetFilterNames(init, GRPC_CLIENT_CHANNEL, ChannelArgs().Set("bar", true)),
+      std::vector<std::string>({"bar", "foo"}));
+  EXPECT_EQ(GetFilterNames(init, GRPC_CLIENT_CHANNEL,
+                           ChannelArgs().Set("bar", true).Set("foo", false)),
+            std::vector<std::string>({"bar"}));
+}
+
+TEST(ChannelInitTest, CanAddTerminalFilter) {
+  ChannelInit::Builder b;
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("foo"));
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("bar")).Terminal();
+  auto init = b.Build();
+  EXPECT_EQ(GetFilterNames(init, GRPC_CLIENT_CHANNEL, ChannelArgs()),
+            std::vector<std::string>({"foo", "bar"}));
+}
+
+TEST(ChannelInitTest, CanAddMultipleTerminalFilters) {
+  ChannelInit::Builder b;
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("foo"));
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("bar"))
+      .Terminal()
+      .IfChannelArg("bar", false);
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("baz"))
+      .Terminal()
+      .IfChannelArg("baz", false);
+  auto init = b.Build();
+  EXPECT_EQ(GetFilterNames(init, GRPC_CLIENT_CHANNEL, ChannelArgs()),
+            std::vector<std::string>({"foo"}));
+  EXPECT_EQ(
+      GetFilterNames(init, GRPC_CLIENT_CHANNEL, ChannelArgs().Set("bar", true)),
+      std::vector<std::string>({"foo", "bar"}));
+  EXPECT_EQ(
+      GetFilterNames(init, GRPC_CLIENT_CHANNEL, ChannelArgs().Set("baz", true)),
+      std::vector<std::string>({"foo", "baz"}));
+}
+
+TEST(ChannelInitTest, CanAddBeforeAllOnce) {
+  ChannelInit::Builder b;
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("foo")).BeforeAll();
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("bar"));
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("baz"));
+  EXPECT_EQ(GetFilterNames(b.Build(), GRPC_CLIENT_CHANNEL, ChannelArgs()),
+            std::vector<std::string>({"foo", "bar", "baz"}));
+}
+
+}  // namespace
```

Comment:
Please also add a test for the post-processor functionality.

Reply:
done